### PR TITLE
Polished TensorFlowVariables code and documentation

### PR DESCRIPTION
**Pull request by Wapaul1**

### CODE REVIEW

File: doc/source/using-ray-with-tensorflow.rst [hunk 1]
```diff
@@ -320,3 +320,38 @@ For reference, the full code is below:
...
+
+.. autoclass:: ray.experimental.TensorFlowVariables
+   :members:
+
+Troubleshooting
+---------------
+
+Note that ``TensorFlowVariables`` uses variable names to determine what variables get set when calling ``set_weights``.
+One common issue when sharing weights between networks is that the setter and getter networks are both defined
+in same Tensorflow graph. This occurs as Tensorflow appends an underscore and integer to the names of variables
```
Comment (pcmoritz) [inline doc/source/using-ray-with-tensorflow.rst]: in _the_ same TensorFlow

File: doc/source/using-ray-with-tensorflow.rst [hunk 2]
```diff
@@ -320,3 +320,38 @@ For reference, the full code is below:
...
     if iteration % 20 == 0:
       print("Iteration {}: weights are {}".format(iteration, weights))
+
+.. autoclass:: ray.experimental.TensorFlowVariables
+   :members:
+
+Troubleshooting
+---------------
+
+Note that ``TensorFlowVariables`` uses variable names to determine what variables get set when calling ``set_weights``.
```
Comment (pcmoritz) [inline doc/source/using-ray-with-tensorflow.rst]: which variables _to_ set

File: doc/source/using-ray-with-tensorflow.rst [hunk 3]
```diff
@@ -320,3 +320,38 @@ For reference, the full code is below:
...
+One common issue when sharing weights between networks is that the setter and getter networks are both defined
```
Comment (pcmoritz) [inline doc/source/using-ray-with-tensorflow.rst]: One common issue arises when two networks are defined in the same TensorFlow graph. In this case, TensorFlow appends an underscore and integer to the names of variables to disambiguate them.

File: doc/source/using-ray-with-tensorflow.rst [hunk 4]
```diff
@@ -320,3 +320,38 @@ For reference, the full code is below:
...
+in same Tensorflow graph. This occurs as Tensorflow appends an underscore and integer to the names of variables
+whose name is already taken in the graph. For example, if I had a class definiton ``Network`` with a ``TensorFlowVariables`` 
+instance inside and ran the following code:
```
Comment (pcmoritz) [inline doc/source/using-ray-with-tensorflow.rst]: get rid of "inside"

File: doc/source/using-ray-with-tensorflow.rst [hunk 5]
```diff
@@ -320,3 +320,38 @@ For reference, the full code is below:
...
+.. autoclass:: ray.experimental.TensorFlowVariables
+   :members:
+
+Troubleshooting
+---------------
+
+Note that ``TensorFlowVariables`` uses variable names to determine what variables get set when calling ``set_weights``.
+One common issue when sharing weights between networks is that the setter and getter networks are both defined
+in same Tensorflow graph. This occurs as Tensorflow appends an underscore and integer to the names of variables
+whose name is already taken in the graph. For example, if I had a class definiton ``Network`` with a ``TensorFlowVariables``
```
Comment (pcmoritz) [inline doc/source/using-ray-with-tensorflow.rst]: replace "I" with "we" everywhere

File: doc/source/using-ray-with-tensorflow.rst [hunk 6]
```diff
@@ -320,3 +320,38 @@ For reference, the full code is below:
...
+  with tf.Graph().as_default():
+    a = Network()
+  with tf.Graph().as_default():
+    b = Network()
+  b.set_weights(a.get_weights())
+
+This issue does not occur between actors that contain a network, as each actor is in its own process, and thus is in
+its own graph. 
+
+Another thing to keep in mind when using the class is that creating a ``TensorFlowVariables`` adds new operations to the graph,
```
Comment (pcmoritz) [inline doc/source/using-ray-with-tensorflow.rst]: Another issue to keep in mind is that ``TensorFlowVariables`` needs to add new operations to the graph. If you close the graph...

File: python/ray/experimental/tfutils.py [hunk 7]
```diff
@@ -91,29 +115,64 @@ def _check_sess(self):
...
     shapes = [v.get_shape().as_list() for v in self.variables.values()]
     arrays = unflatten(new_weights, shapes)
     placeholders = [self.placeholders[k] for k, v in self.variables.items()]
-    self.sess.run(self.assignment_nodes,
+    self.sess.run(self.assignment_nodes.values(),
                   feed_dict=dict(zip(placeholders, arrays)))
 
   def get_weights(self):
-    """Returns the weights of the variables of the loss function in a list."""
+    """Returns the weights in a dictionary.
```
Comment (pcmoritz) [inline python/ray/experimental/tfutils.py]: Returns a dictionary containing the weights of the network

File: python/ray/experimental/tfutils.py [hunk 8]
```diff
@@ -91,29 +115,64 @@ def _check_sess(self):
...
+
+    """
     self._check_sess()
-    self.sess.run(self.assignment_nodes,
+    assign_list = [self.assignment_nodes[name] 
+                   for name in new_weights.keys() 
+                   if name in self.assignment_nodes]
+    assert len(assign_list) != 0, ("No variables in the input matched those in the network. "
+                                   "This could be because this network and another were defined "
+                                   "in the same tensorflow graph. To fix this, place each network "
```
Comment (pcmoritz) [inline python/ray/experimental/tfutils.py]: TensorFlow

File: python/ray/experimental/tfutils.py [hunk 9]
```diff
@@ -91,29 +115,64 @@ def _check_sess(self):
...
+        new_weights (Dict): Dictionary mapping variable names to their weights.
+
+    """
     self._check_sess()
-    self.sess.run(self.assignment_nodes,
+    assign_list = [self.assignment_nodes[name] 
+                   for name in new_weights.keys() 
+                   if name in self.assignment_nodes]
+    assert len(assign_list) != 0, ("No variables in the input matched those in the network. "
+                                   "This could be because this network and another were defined "
```
Comment (pcmoritz) [inline python/ray/experimental/tfutils.py]: Possible cause: Two networks were defined in the same TensorFlow graph.

File: python/ray/experimental/tfutils.py [hunk 10]
```diff
@@ -18,24 +18,31 @@ def unflatten(vector, shapes):
...
+  """A class used to set and get weights for Tensorflow networks.
 
   Attributes:
     sess (tf.Session): The tensorflow session used to run assignment.
-    loss: The loss function passed in by the user.
-    variables (List[tf.Variable]): Extracted variables from the loss.
-    assignment_placeholders (List[tf.placeholders]): The nodes that weights get
-      passed to.
-    assignment_nodes (List[tf.Tensor]): The nodes that assign the weights.
+    variables (List[tf.Variable]): Extracted variables from the loss or passed in variables.
```
Comment (pcmoritz) [inline python/ray/experimental/tfutils.py]: or additional variables that are passed in

File: python/ray/experimental/tfutils.py [hunk 11]
```diff
@@ -18,24 +18,31 @@ def unflatten(vector, shapes):
...
+    placeholders (Dict[str, tf.placeholders]): The nodes that weights get passed to.
```
Comment (pcmoritz) [inline python/ray/experimental/tfutils.py]: Placeholders for the weights.

**Comment by richardliaw** [comment]
@pcmoritz @Wapaul1 are we going to merge these changes?

**Comment by robertnishihara** [comment]
@richardliaw probably should, what do you guys think?

**Comment by richardliaw** [comment]
Sounds good; let me add some comments.

File: python/ray/experimental/tfutils.py [hunk 12]
```diff
@@ -91,29 +115,64 @@ def _check_sess(self):
...
+    Args:
+        new_weights (Dict): Dictionary mapping variable names to their weights.
+
+    """
     self._check_sess()
-    self.sess.run(self.assignment_nodes,
+    assign_list = [self.assignment_nodes[name] 
+                   for name in new_weights.keys() 
+                   if name in self.assignment_nodes]
+    assert len(assign_list) != 0, ("No variables in the input matched those in the network. "
```
Comment (richardliaw) [inline python/ray/experimental/tfutils.py]: nit: can do `assert assign_list, ("No Variables...`

File: test/tensorflow_test.py [hunk 13]
```diff
@@ -125,6 +147,32 @@ def testVariableNameCollision(self):
 
     ray.worker.cleanup()
 
+  # Test that TensorFlowVariables can take in addition variables through
```
Comment (richardliaw) [inline test/tensorflow_test.py]: Can you put this as a docstring inside the test function?

File: test/tensorflow_test.py [hunk 14]
```diff
@@ -125,6 +147,32 @@ def testVariableNameCollision(self):
...
+    net = LossActor(use_loss=False)
+    self.assertEqual(len(net.values[0].variables.items()), 1)
+    self.assertEqual(len(net.values[0].placeholders.items()), 1)
+    
+    net.values[0].set_weights(net.values[0].get_weights())
+
+    ray.worker.cleanup()
+
+  # Test that TensorFlowVariables can take in addition variables through
+  # input_variables arg and with a loss.
```
Comment (richardliaw) [inline test/tensorflow_test.py]: Can you put this as a docstring inside the function?

**Review by richardliaw — COMMENTED** [review]
lgtm after small comments and fix merge conflicts. @Wapaul1

File: python/ray/experimental/tfutils.py [hunk 15]
```diff
@@ -18,104 +18,161 @@ def unflatten(vector, shapes):
...
+  """A class used to set and get weights for Tensorflow networks.
+
+  Attributes:
+    sess (tf.Session): The tensorflow session used to run assignment.
+    variables (Dict[str, tf.Variable]): Extracted variables from the loss or additional variables that are passed in.
+    placeholders (Dict[str, tf.placeholders]): Placeholders for the weights.
+    assignment_nodes (Dict[str, tf.Tensor]): The nodes that assign the weights.
+
+  """
+  def __init__(self, loss, sess=None, input_variables=None):
```
Comment (richardliaw) [inline python/ray/experimental/tfutils.py]: @Wapaul1 can you fix the indentation to 4 spaces?

Comment (richardliaw) [reply python/ray/experimental/tfutils.py]: @Wapaul1 just bumping this

**Comment by robertnishihara** [comment]
@Wapaul1 looks like there are some linting errors.

**Comment by richardliaw** [comment]
this is still not passing lint, but I can fix it later..

**Comment by robertnishihara** [comment]
Thanks @richardliaw, I'm taking a quick look now.

**Comment by robertnishihara** [comment]
This looks good to me, however, I can't quite remember the exact use case we had in mind for this. @Wapaul1 or @richardliaw can you remind me? It'd be great to explain it in the PR description.

**Comment by richardliaw** [comment]
I think this was just cleaning up the docs - no particular necessity

**Comment by Wapaul1** [comment]
Documentation, updates to tests, and allowing users to add extra variables they want to set.

**Comment by robertnishihara** [comment]
Thanks @Wapaul1, I guess what I'm asking is when do users need/want to add extra variables? What applications/scenarios give rise to this?

**Comment by Wapaul1** [comment]
One scenario, and the primary reason I added this functionality, is if you are only training part of a network. If you plug only part of the network into TensorFlowVariables, new nodes will also be created for all of its dependencies via BFS, which would inflate the size of the weights needlessly. This does not occur for variables passed through the new argument.

**Review by robertnishihara — APPROVED** [review]

**Pull request merged**